src/ir_def/check_ir.ml

```diff
@@ -1067,6 +1067,7 @@ and type_exp_field env s f : T.field =
     with Not_found -> error env f.at "field typing for %s not found" name
   in
   assert (t <> T.Pre);
+  let t = T.(match f.note with Mut _ -> t | _ -> as_immut t) in
```

Comment:
This covers the subsumption that an immutable field can be initialised from a mutable variable by just dereferencing the mutcell.